File: exam-3/PlaylistManager-Python/MusicPlaylistManagerApp.py

from Solution import Song, Playlist, PlaylistManager

def test_song():
    print("Running Song Tests...")
    # Test Case 1: Create song and verify details string
    s1 = Song("Imagine", "John Lennon", "Imagine", "Rock", 183)
    details = s1.display_details()
    assert "Imagine" in details, "Test Case 1 Failed: Details should contain the title."
    
    # Test Case 2: Check artist assignment
    assert s1.artist == "John Lennon", "Test Case 2 Failed: Artist is incorrect."
    
    # Test Case 3: Check album assignment
    assert s1.album == "Imagine", "Test Case 3 Failed: Album is incorrect."
    
    # Test Case 4: Check duration
    assert s1.duration == 183, "Test Case 4 Failed: Duration is incorrect."
    
    print("Song Tests Passed.")
    print("----------------------------------------------------")

def test_playlist():
    print("Running Playlist Tests...")
    playlist = Playlist("Favorites")
    
    # Create songs
    s1 = Song("Song A", "Artist1", "Album1", "Pop", 200)
    s2 = Song("Song B", "Artist2", "Album2", "Rock", 240)
    s3 = Song("Song C", "Artist1", "Album3", "Pop", 180)
    s4 = Song("Song D", "Artist3", "Album4", "Jazz", 220)
    
    # Test Case 1: Add songs and check count.
    playlist.add_song(s1)
    playlist.add_song(s2)
    playlist.add_song(s3)
    playlist.add_song(s4)
    assert len(playlist.get_songs()) == 4, "Test Case 1 Failed: There should be 4 songs in the playlist."
    
    # Test Case 2: Remove a song and check count.
    removed = playlist.remove_song("Song B")
    assert removed, "Test Case 2 Failed: Song B should be removed."
    assert len(playlist.get_songs()) == 3, "Test Case 2 Failed: There should be 3 songs after removal."
    
    # Test Case 3: Filter songs by artist "Artist1"
    filtered = playlist.filter_songs("artist", "Artist1")
    assert len(filtered) == 2, "Test Case 3 Failed: There should be 2 songs by Artist1."
    
    # Test Case 4: Search for songs with keyword "Jazz"
    search_results = playlist.search_songs("Jazz")
    assert len(search_results) == 1, "Test Case 4 Failed: There should be 1 song with Jazz in details."
    
    print("Playlist Tests Passed.")
    print("----------------------------------------------------")

def test_playlist_manager():
    print("Running PlaylistManager Tests...")
    manager = PlaylistManager()
    
    # Test Case 1: Create playlists and verify count.
    manager.create_playlist("Chill")
    manager.create_playlist("Workout")
    manager.create_playlist("Party")
    manager.create_playlist("Study")
    assert len(manager.list_playlists()) == 4, "Test Case 1 Failed: There should be 4 playlists."
    
    # Test Case 2: Get a specific playlist and add a song.
    chill = manager.get_playlist("Chill")
    assert chill is not None, "Test Case 2 Failed: Playlist 'Chill' should exist."
    s = Song("Calm Song", "ArtistX", "AlbumX", "Ambient", 300)
    chill.add_song(s)
    assert len(chill.get_songs()) == 1, "Test Case 2 Failed: 'Chill' should have 1 song."
    
    # Test Case 3: Delete a playlist and verify deletion.
    deleted = manager.delete_playlist("Party")
    assert deleted, "Test Case 3 Failed: 'Party' playlist should be deleted."
    assert len(manager.list_playlists()) == 3, "Test Case 3 Failed: There should be 3 playlists remaining."
    
    # Test Case 4: Cross-playlist search.
    workout = manager.get_playlist("Workout")
    workout.add_song(Song("Energy", "ArtistY", "AlbumY", "Rock", 210))
    study = manager.get_playlist("Study")
    study.add_song(Song("Focus", "ArtistZ", "AlbumZ", "Classical", 260))
    cross_results = manager.cross_playlist_search("ArtistY")
    assert len(cross_results) == 1, "Test Case 4 Failed: Cross-search should find 1 song by ArtistY."
    
    print("PlaylistManager Tests Passed.")
    print("----------------------------------------------------")

def integration_test():
    print("Running Extended Integration Tests...")
    manager = PlaylistManager()
    manager.create_playlist("Road Trip")
    manager.create_playlist("Relaxation")
    manager.create_playlist("Workout")
    
    road_trip = manager.get_playlist("Road Trip")
    relaxation = manager.get_playlist("Relaxation")
    workout = manager.get_playlist("Workout")
    
    # Add songs to Road Trip playlist
    road_trip.add_song(Song("Highway Tune", "Greta Van Fleet", "Anthem of the Road", "Rock", 210))
    road_trip.add_song(Song("Life is a Highway", "Tom Cochrane", "Mad Mad World", "Rock", 250))
    road_trip.add_song(Song("Drive", "Incubus", "Make Yourself", "Alternative", 260))
    
    # Add songs to Relaxation playlist
    relaxation.add_song(Song("Ocean Eyes", "Billie Eilish", "Don't Smile", "Pop", 200))
    relaxation.add_song(Song("Weightless", "Marconi Union", "Ambient", "Ambient", 500))
    
    # Add songs to Workout playlist
    workout.add_song(Song("Stronger", "Kanye West", "Graduation", "Hip Hop", 312))
    workout.add_song(Song("Eye of the Tiger", "Survivor", "Eye of the Tiger", "Rock", 245))
    
    # --- Extended Testing Scenarios ---
    
    # Scenario 1: Verify filtering within Road Trip playlist for Rock songs.
    road_trip_rock = road_trip.filter_songs("genre", "Rock")
    assert len(road_trip_rock) == 2, "Integration Test Scenario 1 Failed: Road Trip should have 2 Rock songs."
    
    # Scenario 2: Search for a partial keyword in Road Trip.
    search_drive = road_trip.search_songs("drive")
    assert len(search_drive) == 1, "Integration Test Scenario 2 Failed: Should find 1 song with 'drive' in the details."
    
    # Scenario 3: Cross-playlist search for genre "Rock"
    cross_search = manager.cross_playlist_search("Rock")
    # Expected: 2 songs from Road Trip + 1 from Workout = 3 songs total
    assert len(cross_search) == 3, "Integration Test Scenario 3 Failed: Cross-search for 'Rock' should find 3 songs."
    
    # Scenario 4: Delete a song from Road Trip and verify updated count.
    removed = road_trip.remove_song("Life is a Highway")
    assert removed, "Integration Test Scenario 4 Failed: 'Life is a Highway' should be removed."
    assert len(road_trip.get_songs()) == 2, "Integration Test Scenario 4 Failed: Road Trip should now have 2 songs."
    
    # Scenario 5: Delete the Workout playlist and verify the playlist list.
    deleted_playlist = manager.delete_playlist("Workout")
    assert deleted_playlist, "Integration Test Scenario 5 Failed: Workout playlist should be deleted."
    assert len(manager.list_playlists()) == 2, "Integration Test Scenario 5 Failed: There should be 2 playlists remaining."
    
    # Scenario 6: Create a new Party playlist and add a song.
    manager.create_playlist("Party")
    party = manager.get_playlist("Party")
    party.add_song(Song("Uptown Funk", "Bruno Mars", "Uptown Special", "Funk", 270))
    assert len(party.get_songs()) == 1, "Integration Test Scenario 6 Failed: Party should have 1 song."
    
    print("Extended Integration Tests Passed.")
    print("----------------------------------------------------")

def main():
    print("Welcome to the Music Playlist Manager!")
    test_song()
    test_playlist()
    test_playlist_manager()
    integration_test()
    print("All tests completed successfully.")

if __name__ == "__main__":
    main()

File: exam-3/PlaylistManager-Python/MusicPlaylistManagerApp.java
import java.util.ArrayList;
import java.util.List;

class SongTest {
    public static void runTests() {
        System.out.println("Running Song Tests...");

        // Test Case 1: Create a song and check details string.
        Song s1 = new Song("Imagine", "John Lennon", "Imagine", "Rock", 183);
        String details = s1.displayDetails();
        if (!details.contains("Imagine")) {
            System.out.println("Test Case 1 Failed: Song details should contain the title.");
        } else {
            System.out.println("Test Case 1 Passed.");
        }

        // Test Case 2: Check that artist is correctly assigned.
        if (!s1.artist.equals("John Lennon")) {
            System.out.println("Test Case 2 Failed: Artist name is incorrect.");
        } else {
            System.out.println("Test Case 2 Passed.");
        }

        // Test Case 3: Check album assignment.
        if (!s1.album.equals("Imagine")) {
            System.out.println("Test Case 3 Failed: Album name is incorrect.");
        } else {
            System.out.println("Test Case 3 Passed.");
        }

        // Test Case 4: Check duration is set correctly.
        if (s1.duration != 183) {
            System.out.println("Test Case 4 Failed: Duration is incorrect.");
        } else {
            System.out.println("Test Case 4 Passed.");
        }

        System.out.println("Song Tests Completed.");
        System.out.println("----------------------------------------------------");
    }
}

class PlaylistTest {
    public static void runTests() {
        System.out.println("Running Playlist Tests...");

        Playlist playlist = new Playlist("Favorites");

        // Test Case 1: Add songs and check count.
        Song s1 = new Song("Song A", "Artist1", "Album1", "Pop", 200);
        Song s2 = new Song("Song B", "Artist2", "Album2", "Rock", 240);
        Song s3 = new Song("Song C", "Artist1", "Album3", "Pop", 180);
        Song s4 = new Song("Song D", "Artist3", "Album4", "Jazz", 220);
        playlist.addSong(s1);
        playlist.addSong(s2);
        playlist.addSong(s3);
        playlist.addSong(s4);
        if (playlist.getSongs().size() != 4) {
            System.out.println("Test Case 1 Failed: There should be 4 songs in the playlist.");
        } else {
            System.out.println("Test Case 1 Passed.");
        }

        // Test Case 2: Remove a song and check count.
        boolean removed = playlist.removeSong("Song B");
        if (!removed) {
            System.out.println("Test Case 2 Failed: Song B should be removed.");
        } else if (playlist.getSongs().size() != 3) {
            System.out.println("Test Case 2 Failed: There should be 3 songs left.");
        } else {
            System.out.println("Test Case 2 Passed.");
        }

        // Test Case 3: Filter songs by artist "Artist1".
        List<Song> filteredByArtist = playlist.filterSongs("artist", "Artist1");
        if (filteredByArtist.size() != 2) {
            System.out.println("Test Case 3 Failed: There should be 2 songs by Artist1.");
        } else {
            System.out.println("Test Case 3 Passed.");
        }

        // Test Case 4: Search songs by keyword "Jazz".
        List<Song> searchResults = playlist.searchSongs("Jazz");
        if (searchResults.size() != 1) {
            System.out.println("Test Case 4 Failed: There should be 1 song with Jazz in details.");
        } else {
            System.out.println("Test Case 4 Passed.");
        }

        System.out.println("Playlist Tests Completed.");
        System.out.println("----------------------------------------------------");
    }
}

class PlaylistManagerTest {
    public static void runTests() {
        System.out.println("Running PlaylistManager Tests...");

        PlaylistManager manager = new PlaylistManager();

        // Test Case 1: Create playlists and verify list.
        manager.createPlaylist("Chill");
        manager.createPlaylist("Workout");
        manager.createPlaylist("Party");
        manager.createPlaylist("Study");
        List<String> names = manager.listPlaylists();
        if (names.size() != 4) {
            System.out.println("Test Case 1 Failed: There should be 4 playlists.");
        } else {
            System.out.println("Test Case 1 Passed.");
        }

        // Test Case 2: Get a specific playlist and add a song.
        Playlist chill = manager.getPlaylist("Chill");
        if (chill == null) {
            System.out.println("Test Case 2 Failed: Playlist 'Chill' should exist.");
        } else {
            Song s = new Song("Calm Song", "ArtistX", "AlbumX", "Ambient", 300);
            chill.addSong(s);
            if (chill.getSongs().size() != 1) {
                System.out.println("Test Case 2 Failed: 'Chill' should have 1 song.");
            } else {
                System.out.println("Test Case 2 Passed.");
            }
        }

        // Test Case 3: Delete a playlist and verify deletion.
        boolean deleted = manager.deletePlaylist("Party");
        if (!deleted) {
            System.out.println("Test Case 3 Failed: 'Party' playlist should be deleted.");
        } else if (manager.listPlaylists().size() != 3) {
            System.out.println("Test Case 3 Failed: There should be 3 playlists after deletion.");
        } else {
            System.out.println("Test Case 3 Passed.");
        }

        // Test Case 4: Cross-playlist search.
        // Add songs to different playlists.
        Playlist workout = manager.getPlaylist("Workout");
        workout.addSong(new Song("Energy", "ArtistY", "AlbumY", "Rock", 210));
        Playlist study = manager.getPlaylist("Study");
        study.addSong(new Song("Focus", "ArtistZ", "AlbumZ", "Classical", 260));
        List<Song> crossSearch = manager.crossPlaylistSearch("ArtistY");
        if (crossSearch.size() != 1) {
            System.out.println("Test Case 4 Failed: Cross-search should find 1 song by ArtistY.");
        } else {
            System.out.println("Test Case 4 Passed.");
        }

        System.out.println("PlaylistManager Tests Completed.");
        System.out.println("----------------------------------------------------");
    }
}

class IntegrationTest {
    public static void runIntegrationTests() {
        System.out.println("Running Extended Integration Tests...");

        // Create a new manager and several playlists.
        PlaylistManager manager = new PlaylistManager();
        manager.createPlaylist("Road Trip");
        manager.createPlaylist("Relaxation");
        manager.createPlaylist("Workout");

        // Retrieve playlists.
        Playlist roadTrip = manager.getPlaylist("Road Trip");
        Playlist relaxation = manager.getPlaylist("Relaxation");
        Playlist workout = manager.getPlaylist("Workout");

        // Add songs to Road Trip playlist.
        roadTrip.addSong(new Song("Highway Tune", "Greta Van Fleet", "Anthem of the Road", "Rock", 210));
        roadTrip.addSong(new Song("Life is a Highway", "Tom Cochrane", "Mad Mad World", "Rock", 250));
        roadTrip.addSong(new Song("Drive", "Incubus", "Make Yourself", "Alternative", 260));

        // Add songs to Relaxation playlist.
        relaxation.addSong(new Song("Ocean Eyes", "Billie Eilish", "Don't Smile", "Pop", 200));
        relaxation.addSong(new Song("Weightless", "Marconi Union", "Ambient", "Ambient", 500));

        // Add songs to Workout playlist.
        workout.addSong(new Song("Stronger", "Kanye West", "Graduation", "Hip Hop", 312));
        workout.addSong(new Song("Eye of the Tiger", "Survivor", "Eye of the Tiger", "Rock", 245));

        // Scenario 1: Verify filtering within the Road Trip playlist.
        List<Song> roadTripRock = roadTrip.filterSongs("genre", "Rock");
        if (roadTripRock.size() != 2) {
            System.out.println("Integration Test Scenario 1 Failed: Road Trip should have 2 Rock songs.");
        } else {
            System.out.println("Integration Test Scenario 1 Passed.");
        }

        // Scenario 2: Extended search in Road Trip playlist (by partial keyword).
        List<Song> searchDrive = roadTrip.searchSongs("drive");
        if (searchDrive.size() != 1) {
            System.out.println("Integration Test Scenario 2 Failed: Should find 1 song with 'drive' in the details.");
        } else {
            System.out.println("Integration Test Scenario 2 Passed.");
        }

        // Scenario 3: Cross-playlist search (multiple matches).
        List<Song> crossSearch = manager.crossPlaylistSearch("Rock");
        if (crossSearch.size() != 3) {
            System.out.println("Integration Test Scenario 3 Failed: Cross-search for 'Rock' should find 3 songs.");
        } else {
            System.out.println("Integration Test Scenario 3 Passed.");
        }

        // Scenario 4: Delete a song from a playlist and re-run search/filter tests.
        boolean removed = roadTrip.removeSong("Life is a Highway");
        if (!removed) {
            System.out.println("Integration Test Scenario 4 Failed: 'Life is a Highway' should be removed.");
        } else if (roadTrip.getSongs().size() != 2) {
            System.out.println("Integration Test Scenario 4 Failed: Road Trip should now have 2 songs.");
        } else {
            System.out.println("Integration Test Scenario 4 Passed.");
        }

        // Scenario 5: Delete a playlist and check the overall playlist list.
        boolean deletedPlaylist = manager.deletePlaylist("Workout");
        if (!deletedPlaylist) {
            System.out.println("Integration Test Scenario 5 Failed: Workout playlist should be deleted.");
        } else if (manager.listPlaylists().size() != 2) {
            System.out.println("Integration Test Scenario 5 Failed: There should be 2 playlists remaining.");
        } else {
            System.out.println("Integration Test Scenario 5 Passed.");
        }

        // Scenario 6: Verify addition of songs to a new playlist after deletion.
        manager.createPlaylist("Party");
        Playlist party = manager.getPlaylist("Party");
        party.addSong(new Song("Uptown Funk", "Bruno Mars", "Uptown Special", "Funk", 270));
        if (party.getSongs().size() != 1) {
            System.out.println("Integration Test Scenario 6 Failed: Party should have 1 song.");
        } else {
            System.out.println("Integration Test Scenario 6 Passed.");
        }

        System.out.println("Integration Tests Completed.");
        System.out.println("----------------------------------------------------");
    }
}

// -----------------------
// Main Application Class
// -----------------------

public class MusicPlaylistManagerApp {
    public static void main(String[] args) {
        System.out.println("Welcome to the Music Playlist Manager!");

        // Run individual tests
        SongTest.runTests();
        PlaylistTest.runTests();
        PlaylistManagerTest.runTests();

        // Extended integration testing
        System.out.println("Running Complete Extended Integration Test...");
        IntegrationTest.runIntegrationTests();

        System.out.println("All tests completed successfully.");
    }
}